shared/src/commonMain/kotlin/screens/landing/LandingScreenViewModel.kt

```diff
@@ -1,19 +1,21 @@
 package screens.landing
 
-import com.rickclephas.kmm.viewmodel.KMMViewModel
 import com.rickclephas.kmm.viewmodel.coroutineScope
+import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.launch
-import screens.IViewModel
+import screens.AbstractViewModel
 
-class LandingScreenViewModel : KMMViewModel(),
```

Comment:
Here is our BaseComposeViewModel we created for ABB:

package com.abbemobility.shared.core.common.base

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.vectorResource
import androidx.compose.ui.unit.dp
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.abbemobility.shared.core.common.R
import com.abbemobility.shared.core.common.alerts.DialogAlert
import com.abbemobility.shared.core.common.base.dialogstate.AlertMessage
import com.abbemobility.shared.core.common.base.dialogstate.EmptyAlertMessage
import com.abbemobility.shared.core.common.base.snackbar.ErrorSnackbarAlert
import com.abbemobility.shared.core.common.base.snackbar.SnackbarAlert
import com.abbemobility.shared.core.common.base.snackbar.SuccessSnackbarAlert
import com.abbemobility.shared.core.common.errorhandling.ErrorType
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.plus

@Suppress("TooManyFunctions")
abstract class BaseComposeViewModel<UIEvent, UIState, Action>(initialStateValue: UIState) : ViewModel() {
    private val _uiState = MutableStateFlow(initialStateValue)
    val uiState = _uiState.asStateFlow()

    val eventFlow = MutableSharedFlow<UIEvent>(replay = 1)

    @Suppress("MemberVisibilityCanBePrivate")
    var vmScope = viewModelScope

    private var snackbarMessageState by mutableStateOf<SnackbarMessage>(Empty)
    private var alertMessageState by mutableStateOf<AlertMessage>(EmptyAlertMessage)

    private val _loading = MutableStateFlow(false)
    val loading = _loading.asStateFlow()

    private val _saving = MutableStateFlow(false)
    val saving = _saving.asStateFlow()

    init {
        setDefaultExceptionHandler()
    }
    
    fun isLoading(loading: Boolean) {
        _loading.tryEmit(loading)
    }

    fun isSaving(saving: Boolean) {
        _saving.tryEmit(saving)
    }

    open fun vmScopeWithCustomExceptionHandler(onError: (error: Throwable) -> Unit = {}) =
        (viewModelScope + CoroutineExceptionHandler { _, throwable ->
            isLoading(false)
            onError(throwable)
            snackbarMessageState = SnackbarMessage(throwable.message)
        })

    open fun launchWithErrorHandling(
        errorType: ErrorType,
        block: suspend CoroutineScope.() -> Unit
    ) = vmScope.launch {
        block()
    }

    fun launchAndLoad(block: suspend CoroutineScope.() -> Unit) = vmScope.launch {
        isLoading(true)
        block()
        isLoading(false)
    }

    fun launchAndSave(block: suspend CoroutineScope.() -> Unit) = vmScope.launch {
        isSaving(true)
        block()
        isSaving(false)
    }

    abstract fun onAction(action: Action)

    fun updateUIState(block: (UIState) -> UIState) {
        _uiState.update { block(it) }
    }

    fun triggerUIEvent(event: UIEvent) {
        eventFlow.tryEmit(event)
    }

    private fun setDefaultExceptionHandler(onError: (error: Throwable) -> Unit = {}) {
        vmScope = vmScopeWithCustomExceptionHandler(onError)
    }

    fun showSnackbar(snackbarMessage: SnackbarMessage) = viewModelScope.launch {
        snackbarMessageState = snackbarMessage
    }

    fun showSnackbar(message: String?) = viewModelScope.launch {
        snackbarMessageState = SnackbarMessage(message = message ?: return@launch)
    }

    fun showDialog(alertMessage: AlertMessage) = viewModelScope.launch {
        alertMessageState = alertMessage
    }

    fun showDialog(title: String, description: String?) = viewModelScope.launch {
        alertMessageState = AlertMessage(
            title = title,
            message = description,
        )
    }


    @Composable
    fun CreateSnackBarHost() {
        val snackBarHostState = remember {
            SnackbarHostState()
        }

        val context = LocalContext.current

        LaunchedEffect(key1 = snackbarMessageState) {
            if (snackbarMessageState !is Empty) {
                snackBarHostState.showSnackbar(snackbarMessageState.getMessage(context) ?: "Unknown error")
                snackbarMessageState = Empty
            }
        }

        SnackbarHost(hostState = snackBarHostState) {
            val snackbarMessage by remember(snackbarMessageState) {
                mutableStateOf(snackbarMessageState)
            }

            val message by remember {
                derivedStateOf {
                    snackbarMessage.getMessage(context).orEmpty()
                }
            }

            when (snackbarMessage) {
                is Success -> SuccessSnackbarAlert(message = message)
                is Error -> ErrorSnackbarAlert(message = message)
                else -> {
                    SnackbarAlert(
                        // This padding is also used by the Snackbar component from Material3
                        modifier = Modifier.padding(12.dp),
                        message = message,
                        icon = snackbarMessage.iconRes?.let { res ->
                            ImageVector.vectorResource(id = res)
                        } ?: ImageVector.vectorResource(id = R.drawable.ic_warning),
                        iconColor = Color.Unspecified,
                        containerColor = snackbarMessage.containerColorRes?.let { res ->
                            colorResource(id = res)
                        } ?: MaterialTheme.colorScheme.surfaceVariant
                    )
                }
            }
        }
    }

    @Composable
    fun CreateDialogAlertHost(
        modifier: Modifier = Modifier,
        content: @Composable () -> Unit
    ) {
        val context = LocalContext.current

        var state by remember { mutableStateOf<AlertMessage>(EmptyAlertMessage) }

        LaunchedEffect(alertMessageState) {
            if(alertMessageState !is EmptyAlertMessage) {
                state = alertMessageState
                alertMessageState = EmptyAlertMessage
            }
        }

        Box(
            modifier = modifier,
        ) {
            content()

            if(state !is EmptyAlertMessage) {
                DialogAlert(
                    title = state.getTitle(context).orEmpty(),
                    description = state.getMessage(context),
                    confirmText = state.getPositiveButton(context).orEmpty(),
                    dismissText = state.getNegativeButton(context),
                    onConfirm = {
                        state.onConfirm()
                        state = EmptyAlertMessage
                    },
                    onDismiss = {
                        state.onDismiss()
                        state = EmptyAlertMessage
                    }
                )
            }
        }
    }
}

Reply:
I took some of it and added it to the custom one made here. I am unsure if I need to take other things from this as well